### PR TITLE
Learning Rate Scheduling

### PR DESCRIPTION
**Pull request by stephenroller**
Warmups are particularly necessary for distributed settings.

Other scheduler settings might help with some of Sam's opinions on TorchAgent.

### CODE REVIEW

File: parlai/core/torch_agent.py [hunk 1]
```diff
@@ -303,9 +328,50 @@ def init_optim(self, params, optim_states=None, saved_optim_type=None):
...
+                verbose=True
+            )
+        elif self.opt.get('lr_scheduler') == 'fixed':
+            self.scheduler = optim.lr_scheduler.StepLR(
+                self.optimizer,
+                patience,
+                gamma=decay,
+            )
+
+    def receive_training_metrics(self, metrics_dict):
```
Comment (stephenroller) [inline parlai/core/torch_agent.py]: todo: delete

File: parlai/core/torch_agent.py [hunk 2]
```diff
@@ -167,6 +167,32 @@ def add_cmdline_args(cls, argparser):
...
+            help='LR scheduler patience. In number of validation runs.'
+        )
+        agent.add_argument(
+            '--lr-scheduler-decay', type=float, default=0.5,
+            help='Decay factor for LR scheduler.'
+        )
+        agent.add_argument(
+            '--warmup-updates', type=int, default=-1,
+            help='Learning rate warmup period. Linearly scales up LR over period.'
+                 'Only enabled if > 0.'
```
Comment (alexholdenmiller) [inline parlai/core/torch_agent.py]: should specify the unit here, e.g. "sgd updates"

File: parlai/core/torch_agent.py [hunk 3]
```diff
@@ -167,6 +167,32 @@ def add_cmdline_args(cls, argparser):
...
+            '--lr-scheduler', type=str, default='reduceonplateau',
+            choices=['reduceonplateau', 'none', 'fixed'],
+            help='Learning rate scheduler.'
+        )
+        agent.add_argument(
+            '--lr-scheduler-patience', type=int, default=3,
+            help='LR scheduler patience. In number of validation runs.'
+        )
+        agent.add_argument(
+            '--lr-scheduler-decay', type=float, default=0.5,
```
Comment (emilydinan) [inline parlai/core/torch_agent.py]: check this is <1, if not, set scheduler to none

File: parlai/core/torch_agent.py [hunk 4]
```diff
@@ -189,6 +191,32 @@ def add_cmdline_args(cls, argparser):
...
+        )
+        agent.add_argument(
+            '--warmup-updates', type=int, default=-1,
+            help='Learning rate warmup period, in number of SGD updates. '
+                 'Linearly scales up LR over period. Only enabled if > 0.'
+        )
+        agent.add_argument(
+            '--warmup-rate', type=float, default=1e-4,
+            help='Warmup learning rate *multiplier*. Initial LR is multiplied by '
+                 'this value. Linearly addjusted up to 1.0 across --warmup-updates '
```
Comment (stephenroller) [inline parlai/core/torch_agent.py]: spelling
